### PR TITLE
Implement routing mechanism 

### PR DESCRIPTION
**Pull request by muzykantov**
- Rename ChainCode -> to Chaincode;
- Implement contract package with Router interface;
- Implement alternate routing mechanism;
- Implement default reflectx.Router;
- Remove unused methods and functions;
- Rethinked Invocation flow;
- Fixed various bugs and code style issues;
- Refactored the code;

Next steps:
- Add more tests;
- Increase new packages coverage;
- Refactor BaseContractInterface;

### CODE REVIEW

File: core/core.go [hunk 1]
```diff
@@ -816,41 +815,49 @@ func (cc *ChainCode) callMethod(
...
 	if err != nil {
 		return nil, err
 	}
 
-	if len(result) < 1 || len(result) > 2 {
-		msg := fmt.Sprintf("invalid number of return values: %d", len(result))
-		span.SetStatus(codes.Error, msg)
-		return nil, errors.New(msg)
+	if len(result) != ep.NumReturns {
+		return nil, fmt.Errorf("expected %d return values, got %d", ep.NumReturns, len(result))
```
Comment (muzykantov) [inline core/core.go:824]: span.SetStatus(codes.Error, msg)

File: core/base_contract.go [hunk 2]
```diff
@@ -207,7 +211,11 @@ func (bc *BaseContract) TxHealthCheck(_ *types.Sender) error {
...
 }
 
-func (bc *BaseContract) GetID() string {
+func (bc *BaseContract) ID() string {
+	return bc.GetID()
+}
+
+func (bc *BaseContract) GetID() string { // deprecated
 	return bc.config.GetSymbol()
 }
```
Comment (pfi79) [inline core/base_contract.go:220]: Why don't you do it the other way around? What's deprecated, run through what's new?

Comment (muzykantov) [reply core/base_contract.go:220]: Good point!

File: core/contract/router.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+// MethodType represents the type of a method in the contract.
+type MethodType int
+
+// Constants representing the different types of methods.
+const (
+	MethodTypeTransaction MethodType = iota
+	MethodTypeInvoke
+	MethodTypeQuery
+)
```
Comment (muzykantov) [inline core/contract/router.go:11]: TODO: Add a comment about backward compatibility.

**Review by pfi79 — APPROVED** [review]

**Pull request merged**